### PR TITLE
Sx0 in adjoint preequilibration

### PR DESCRIPTION
**Pull request by paulstapor**
If adjoint preequllibraiton was not solved algebraically but via integration, there may be a contribution from sx0 * xB, as xB may be != 0 for singular Jacobians... So we may get a non-trivial contribution here. This PR tries to fix that...

### CODE REVIEW

File: include/amici/steadystateproblem.h [hunk 1]
```diff
@@ -321,6 +320,12 @@ class SteadystateProblem {
      */
     void getAdjointUpdates(Model &model, const ExpData &edata);
 
+    /**
+     * @brief Accessor for xB
+     * @return xB
```
Comment (dweindl) [inline include/amici/steadystateproblem.h:325]:
```suggestion
     * @brief Return the adjoint state
     * @return xB adjoint state
```

File: src/rdata.cpp [hunk 2]
```diff
@@ -444,6 +445,21 @@ void ReturnData::handleSx0Backward(const Model &model,
     auto xQBpreeq = preeq.getAdjointQuadrature();
     for (int ip = 0; ip < model.nplist(); ++ip)
         xQB[ip] += xQBpreeq[ip];
+
+    /* Prefer explicit type declaration here, since this MUST be a reference
+     * (for memory reasons), rather than passing this thing by value */
+    const AmiVectorArray &sx0preeq = preeq.getStateSensitivity();
+    auto xBpreeq = preeq.getAdjointState();
```
Comment (dweindl) [inline src/rdata.cpp:452]:
```suggestion
    const auto& xBpreeq = preeq.getAdjointState();
```
(or is there a reason for copying?)

File: src/rdata.cpp [hunk 3]
```diff
@@ -444,6 +445,21 @@ void ReturnData::handleSx0Backward(const Model &model,
     auto xQBpreeq = preeq.getAdjointQuadrature();
     for (int ip = 0; ip < model.nplist(); ++ip)
         xQB[ip] += xQBpreeq[ip];
+
+    /* Prefer explicit type declaration here, since this MUST be a reference
+     * (for memory reasons), rather than passing this thing by value */
```
Comment (dweindl) [inline src/rdata.cpp:450]: Not sure if more helpful or confusing. Just taking the reference here would be clear enough for me.

File: src/rdata.cpp [hunk 4]
```diff
@@ -444,6 +445,21 @@ void ReturnData::handleSx0Backward(const Model &model,
...
+    const AmiVectorArray &sx0preeq = preeq.getStateSensitivity();
+    auto xBpreeq = preeq.getAdjointState();
+
+    /* Add the contribution for sx0 from preequilibration. If backward
+     * preequilibration was done by simulation due to a singular Jacobian,
+     * xB does not need to be 0 and we get a non-zero contribution here. */
```
Comment (dweindl) [inline src/rdata.cpp:456]:
```suggestion
     * xB is not necessarily 0 and we get a non-zero contribution here. */
```

**Review by dweindl — APPROVED** [review]
Good as far as I can tell. Can we easily add a test cases for that using one of the existing models?

File: src/rdata.cpp (hunk 3 above)
Comment (paulstapor) [reply src/rdata.cpp:450]: changed it

**Comment by paulstapor** [comment]
> Good as far as I can tell. Can we easily add a test cases for that using one of the existing models?

Not as far as I know... There's no model which would be easily usable for this. Either I take an existing one an modify it on the fly using libsbml and generate the PEtab files for it... Or I rather add another test case with a PEtab model... The test case we want to cover is based on PEtab, so I'd suggest an integration test with a PEtab model here... What do you think? Too complex?

**Comment by paulstapor** [comment]
Maybe one more thing:
I'd prefer to add a test on this when fixing #1156, as this one also needs to be covered then. Moreover, we will need this fix here in CanPathPro asap. As I anyway want to finish up the steadystate solver issues as soon as I'll find time (in latest 1-2 months), I'd suggest to add a more comprehensive test later, as this will be needed anyways then...

**Comment by dweindl** [comment]
> I'd suggest to add a more comprehensive test later, as this will be needed anyways then...

works for me

**Pull request merged**